Update example notebooks to use xarray

w-k-jones opened the pull request:
I've started work on updating the example notebooks to save/load data using xarray for the v1.6 release. As part of this I thought I would go through and improve some of the descriptions and visualisations in the notebooks.

Note that the plotting functions have not been updated to use xarray. I've added decorators for iris conversion to handle this for now, but there are some issues with time conversions.

* [ ] Have you followed our guidelines in CONTRIBUTING.md? 
* [ ] Have you self-reviewed your code and corrected any misspellings? 
* [ ] Have you written documentation that is easy to understand?
* [ ] Have you written descriptive commit messages? 
* [ ] Have you added NumPy docstrings for newly added functions? 
* [ ] Have you formatted your code using black? 
* [ ] If you have introduced a new functionality, have you added adequate unit tests?
* [ ] Have all tests passed in your local clone? 
* [ ] If you have introduced a new functionality, have you added an example notebook?
* [ ] Have you kept your pull request small and limited so that it is easy to review? 
* [ ] Have the newest changes from this branch been merged?

w-k-jones commented:
@freemansw1 @JuliaKukulies @fsenf I've only made updates to one of the notebooks so far. Let me know if you think these changes are good and I'll replicate them across the other notebooks as well

I intend to keep a duplicate of one/two of the notebooks as they were previously as a "legacy" example to demonstrate that iris still works

w-k-jones commented:
> > @freemansw1 @JuliaKukulies @fsenf I've only made updates to one of the notebooks so far. Let me know if you think these changes are good and I'll replicate them across the other notebooks as well
> > I intend to keep a duplicate of one/two of the notebooks as they were previously as a "legacy" example to demonstrate that iris still works
> 
> Thanks @w-k-jones . I'm happy with the changes. What issues specifically are you having with the plotting functions with time?

`tobac.animation_mask_field` fails to plot the masks/tracks for some frames in a semi-regular pattern (5,7,14,16,23,25...) even though features for those frames are present, so I expect it is some mis-match in the converted timestamps. It's not a blocking issue

w-k-jones commented:
@fsenf @freemansw1 This PR is now ready for review, and all the changes from #489 have been merged. The failing test is the code coverage requirement, which is low because of the changes to the plotting module which is entirely untested. I don't think it's worth adding tests for this, as we plan to reduce support for this part of tobac anyway

fsenf reviewed (approved):
Great Update.

A minor style improvement could be in 
- Example_OLR_Tracking_satellite
- Example_OLR_Tracking_model
- Example_Precip_Tracking

The cell with e.g. (and similar)
```python
Precip[frame].plot(cmap="Blues")
Mask[frame].where(Mask[frame]>0).plot(cmap="tab20")
```
leads to
1. wrongly labeled colorbars
2. partially overlapping colorbars

freemansw1 commented on tobac/plotting.py:
```diff
@@ -26,6 +26,8 @@
+from tobac.utils.internal.coordinates import find_dataframe_horizontal_coords
 
 
 def plot_tracks_mask_field_loop(
```
We might just throw a warning/error here if xarray is passed into these functions.

freemansw1 reviewed (approved):
Thanks, @w-k-jones . I am happy for this to be merged and for v1.6.0 to be released. Some notes on oddities that we should resolve down the line:

- `get_spacings` should probably be retired. It's causing headaches, and frankly most users of tobac should simply know their grid spacing. 
- The feature detection parameter examples haven't been converted to xarray. I can make that my problem for a v1.6.1, since I wrote most of those.

w-k-jones commented:
@fsenf @freemansw1 thanks for the feedback. I've updated the remainder of the notebooks (and changed the corresponding docs from referring to iris cubes -> xarray dataarrays) and fixed the overlapping colorbars. I also realised using `xr.zeros_like` to generate the segmentation mask from the input field was copying over the attributes (e.g. units and standard name) so I changed segmentation to drop these attributes

Regarding the plotting functions, we could add the `xarray_to_iris` decorators to all of them but I'm not sure if this is needed too much. Maybe if someone complains. And I agree about `get_spacings`, and I think long term we want to move away from needing a specific grid spacing if the input field has geospatial coordinates

w-k-jones commented:
Can we drop Python 3.8 support for v1.6? 🙃

fsenf commented:
@w-k-jones : Thank you for the update! Colorbar labels are still wrong. Do you need to rerun the notebooks with your latest commit?

w-k-jones commented:
> @w-k-jones : Thank you for the update! Colorbar labels are still wrong. Do you need to rerun the notebooks with your latest commit?

Ah seems to be this bug: https://github.com/pydata/xarray/issues/10027

I'll see if there's a nice way of handling it without needing the most recent version of xarray

w-k-jones commented:
@fsenf The colourbar labels should now be fixed, as well as the issue with Python 3.8

fsenf commented:
> @fsenf The colourbar labels should now be fixed, as well as the issue with Python 3.8

Perfect! I'm happy if you continue with the merge.
